scipy/stats/tests/test_stats.py

```diff
@@ -7917,6 +7917,140 @@ def test_alias(self):
+
+    @pytest.mark.parametrize(
+        'p, alpha, lb, ub, alternative',
+        [[0.3, 0.95, 1.221402758160170, 1.476980793882643, 'two-sided'],
+         [0.5, 0.9, 1.506817785112854, 1.803988415397857,'two-sided'],
+         [0.25, 0.95, -np.inf, 1.39096812846378, 'less'],
+         [0.8, 0.9, 2.117000016612675, np.inf, 'greater']]
+    )
+    def test_R_ci_quantile(self, p, alpha, lb, ub, alternative):
+        # Test against R library `confintr` function `ci_quantile`, e.g.
+        # library(confintr)
+        # options(digits=16)
+        # x <- exp(seq(0, 1, by = 0.01))
+        # ci_quantile(x, q = 0.3)$interval
+        # ci_quantile(x, q = 0.5, probs = c(0.05, 0.95))$interval
+        # ci_quantile(x, q = 0.25, probs = c(0, 0.95))$interval
+        # ci_quantile(x, q = 0.8, probs = c(0.1, 1))$interval
+        x = np.exp(np.arange(0, 1.01, 0.01))
+        res = stats.quantile_test(x, p=p, alternative=alternative)
+        assert_allclose(res.confidence_interval(alpha), [lb, ub], rtol=1e-15)
+
+    @pytest.mark.parametrize(
+        'q, p, alternative, ref',
+        [[1.2, 0.3, 'two-sided', 0.01515567517648],
+         [1.8, 0.5, 'two-sided', 0.1109183496606]]
+    )
+    def test_R_pvalue(self, q, p, alternative, ref):
+        # Test against R library `snpar` function `quant.test`, e.g.
+        # library(snpar)
+        # options(digits=16)
+        # x < - exp(seq(0, 1, by=0.01))
+        # quant.test(x, q=1.2, p=0.3, exact=TRUE, alternative='t')
+        x = np.exp(np.arange(0, 1.01, 0.01))
+        res = stats.quantile_test(x, q=q, p=p, alternative=alternative)
+        assert_allclose(res.pvalue, ref, rtol=1e-12)
+
+    @pytest.mark.parametrize('alternative', ['less', 'greater'])
+    @pytest.mark.parametrize('alpha', [0.9, 0.95])
+    def test_pval_ci_match(self, alternative, alpha):
+        # Verify that the following statement holds:
+        #
+        # The 95% confidence interval corresponding with alternative='less' 
+        # has -inf as its lower bound, and the upper bound `xu` is the greatest
+        # element from the sample `x` such that:
+        # `stats.quantile_test(x, q=xu, p=p, alternative='less').pvalue``
+        # will be greater than 5%.
+        # 
+        # And the corresponding statement for the alternative='greater' case.
+
+        rng = np.random.default_rng(7716774463289760271)
+        p, q = rng.random(size=2)
+        x = np.sort(rng.random(size=100))
+
+        res = stats.quantile_test(x, q=q, p=p, alternative=alternative)
+        lb, ub = res.confidence_interval(confidence_level=alpha)
+
+        # Interpretation for `alternative='less'`: the upper bound is the
+        # largest element from the sample that is still *inside* the CI.
+        q_inside = ub if alternative == 'less' else lb
+        i = np.where(x == q_inside)[0][0]
+        # The next element from the sample that is greater is *outside* the CI.
+        q_outside = x[i+1] if alternative == 'less' else x[i-1]
```

Comment:
One would also need to change the definition of `q_outside` to be "the next sample value larger/smaller than q_inside"

Reply:
Putting it all together, I get something like that (that passes the current test)
```python
# crafting a sample with a CI that fall on repeated observations
# AND for which we have sample values above/below the CI bounds
x = [0.4]*10 + [0.6]*40 + [0.8]*(10) + [1]*(100-10-40-10)

res = stats.quantile_test(x, q=q, p=p, alternative=alternative)
lb, ub = res.confidence_interval(confidence_level=alpha)

# Interpretation for `alternative='less'`: the upper bound is the
# largest element from the sample that is still *inside* the CI.
q_inside = ub if alternative == 'less' else lb

n = len(x)
bd = stats.binom(n, p)
i = int(bd.isf(1-alpha)) if alternative == 'less' else (int(bd.ppf(1-alpha)) - 1)
q_outside = q_inside
while (q_outside == q_inside):
    i = i+1 if alternative == 'less' else i-1
    q_outside = x[i] 

# .. the rest remains unchanged
```

What do you think?

Reply:
> observation. Essentially, we must replace... What do you think?

I will think through it shortly. In the meantime, it would be helpful if you were to include *why* `where` is wrong and `isf` is right.

Reply:
Sure. 

`where` returns all indexes that match the condition. By taking `[0][0]` we select the smallest one. For example, 

```python
rng = np.random.default_rng(7716774463289760271)
p, q = rng.random(size=2)
x = [0.4]*10 + [0.6]*40 + [0.8]*(10) + [1]*(100-10-40-10)
res = stats.quantile_test(x, q=q, p=p, alternative='less')
lb, ub = res.confidence_interval(confidence_level=alpha)
q_inside = ub if alternative == 'less' else lb
i = np.where(x == q_inside)[0][0]

print(np.where(x == q_inside))
> (array([50, 51, 52, 53, 54, 55, 56, 57, 58, 59]),)
print(i)
> 50
```
The CI bound value is `0.8` and `x[50] = 0.8`. The problem is that the CI bound _index_ isn't `50`. It's `55`. This is the value returned by `isf` (that's how it is implemented in the  `confidence_interval` method.

So, with the current method of computing the `i`, the `q_inside` is correct, and therefore this test passes
```python
res = stats.quantile_test(x, q=q_inside, p=p, alternative=alternative)
assert res.pvalue > 1-alpha
```
but the value of `q_outside` is wrong. We get `q_outside = x[51] = 0.8` and therefore this test fails
 ```python
res = stats.quantile_test(x, q=q_outside, p=p, alternative=alternative)
assert res.pvalue <= 1-alpha
```
What we need to test instead is `q_outside = x[60] = 1` in this case. And with that, the test passes as expected.